### PR TITLE
Advanced Feature: Class Loading

### PR DESCRIPTION
**Pull request by awholegunch**
Takes the blog post on class loading and adds it as an advanced features section. 

Fix for https://issues.cask.co/browse/CDAP-2930.

Staged at: [Java Class Loading and Distributed Data Processing Frameworks](http://docs-staging.cask.co/staging/cdap/3.2.0-SNAPSHOT-feature-3.2_classloading/en/developers-manual/advanced/class-loading.html)

### CODE REVIEW

**Comment by awholegunch** [comment]
"On-hold" as revisions needed to focus the page on the CDAP-related aspects of class loading.

**Comment by awholegunch** [comment]
Revised and ready for re-review.

File: cdap-docs/developers-manual/source/advanced/class-loading.rst [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+    :author: Cask Data, Inc.
+    :copyright: Copyright © 2015 Cask Data, Inc.
+
+.. _class-loading:
+
+=============================================================
+Java Class Loading and Distributed Data Processing Frameworks
+=============================================================
+
+Java class loading is one of the most fundamental and powerful concepts provided by the
```
Comment (bdmogal) [inline cdap-docs/developers-manual/source/advanced/class-loading.rst]: Should we remove this first paragraph? Also replace the heading on line 8 with 'Overview/Introduction'? I think this paragraph is excellent for a blog, but less suitable for product documentation?

File: cdap-docs/developers-manual/source/advanced/class-loading.rst [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+class loading strategy, we provide these properties for an application:
+
+- Application can choose to use any library of any version;
+- Application can define a custom plugin API for extending itself;
+- Plugins can be written for an application by implementing the application plugin API; and
+- Different plugins can choose to use any library of any version, yet all plugins are
+  usable by the same application at the same time.
+
+
+Parent Delegation Model
```
Comment (bdmogal) [inline cdap-docs/developers-manual/source/advanced/class-loading.rst]: Should we move this to a a separate section and just add a reference to it? I feel documentation of Classloading in CDAP should have more emphasis on this page, and it starts from the next section onwards.

File: cdap-docs/developers-manual/source/advanced/class-loading.rst [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+Inside an application bundle JAR file, besides the class files for the application, are
+the JAR files for third party libraries that are needed by the application. In the
+application JAR manifest, a special attribute (“Export-Package”) can be added to instruct
+the filter class loader which packages are contained in the application plugin API. Hence,
+only classes under those packages will be available to the plugin. This is to avoid
+libraries used by the application interfering with the ones used by the plugin. The plugin
+JAR file is following a similar structure as an application JAR file.
+
+Class Loading in Hadoop MapReduce and Apache Spark
```
Comment (bdmogal) [inline cdap-docs/developers-manual/source/advanced/class-loading.rst]: seems a little inconsistent to use Apache for Spark but not for Hadoop. Maybe use Apache Hadoop MapReduce and Apache Spark in the title, and only MapReduce and Spark in the content below.

File: cdap-docs/developers-manual/source/index.rst [hunk 4]
```diff
@@ -68,6 +68,6 @@ CDAP Developers’ Manual
 .. _advanced: advanced/index.html
 
 - |advanced|_ Covers **advanced topics on CDAP** that will be of interest to
-  developers who want a deeper dive into CDAP, with presentations on suggested **best
-  practices for CDAP development,** and on **adding a custom logback** to a CDAP
-  application.
+  developers who want a deeper dive into CDAP, with presentations on **class loading in
```
Comment (bdmogal) [inline cdap-docs/developers-manual/source/index.rst]: maybe another word for presentations?

**Comment by bdmogal** [comment]
2 comments, otherwise LGTM

**Comment by awholegunch** [comment]
@bdmogal I've updated based on your comments. Can you take a final look?

File: cdap-docs/developers-manual/source/advanced/class-loading.rst [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+Inside an application bundle JAR file, besides the class files for the application, are
+the JAR files for third party libraries that are needed by the application. In the
+application JAR manifest, a special attribute (“Export-Package”) can be added to instruct
+the filter class loader which packages are contained in the application plugin API. Hence,
+only classes under those packages will be available to the plugin. This is to avoid
+libraries used by the application interfering with the ones used by the plugin. The plugin
+JAR file is following a similar structure as an application JAR file.
+
+Class Loading in Apache Hadoop MapReduce and Spark
```
Comment (bdmogal) [inline cdap-docs/developers-manual/source/advanced/class-loading.rst]: Apache Spark?

**Comment by bdmogal** [comment]
minor comment, LGTM otherwise

File: cdap-docs/developers-manual/source/advanced/class-loading.rst (hunk 5 above)
Comment (awholegunch) [inline cdap-docs/developers-manual/source/advanced/class-loading.rst]: Sure.

File: cdap-docs/developers-manual/source/advanced/class-loading.rst [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+=====================
+Class Loading in CDAP
+=====================
+
+In this advanced section, we talk about how we used Java class loading to design an
+extensible application framework with a flexible class loading strategy. In terms of class
+loading strategy, we provide these properties for an application:
+
+- Application can choose to use any library of any version;
```
Comment (anew) [inline cdap-docs/developers-manual/source/advanced/class-loading.rst]: An Application?

File: cdap-docs/developers-manual/source/advanced/class-loading.rst [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+- Application can define a custom plugin API for extending itself;
```
Comment (anew) [inline cdap-docs/developers-manual/source/advanced/class-loading.rst]: Ditto

File: cdap-docs/developers-manual/source/advanced/class-loading.rst [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+<https://www-01.ibm.com/support/knowledgecenter/#!/SSYKE2_7.0.0/com.ibm.java.zos.70.doc/diag/understanding/cl_delegation.html>`__. 
+The parent delegation model allows for one class to be loadable from multiple class
+loaders, as long as those class loaders have the defining class loader of that class as a
+common ancestor. This property enables one to define an extensible class loading
+architecture that provides class isolation, yet at the same time they’re able to
+inter-operate with a set of common API classes.
+
+Class Loading in CDAP
+=====================
+Based on the parent delegation model, we’ve come up with this class loader hierarchy in
```
Comment (anew) [inline cdap-docs/developers-manual/source/advanced/class-loading.rst]: The CDAP class loader hierarchy is based on the parent delegation model:

File: cdap-docs/developers-manual/source/advanced/class-loading.rst [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+When launching a MapReduce or Spark job from CDAP, we want to provide the same class
+isolation behavior as described in the previous section. To do so, we alter the classpath
+of the YARN container for the job and make it first call the main method of our launcher
+class. The launcher class will setup the class loader hierarchy as desired and then
+delegate the call to the actual MapReduce or Spark task runner.
+
+Summary
+=======
+
+Java class loading is a powerful and important concept provided by the Java platform.
```
Comment (anew) [inline cdap-docs/developers-manual/source/advanced/class-loading.rst]: I don't think we need this summary. This is juts a chapter in the dev guide, not a blog or article that strands by itself.

**Pull request merged**